src/listener_registry.rs

```diff
@@ -0,0 +1,240 @@
+        // [impl->dsn~utransport-registerlistener-error-resource-exhausted~1]
+        if locked_subscribers.len() >= self.max_subscribers {
+            return Err(UStatus::fail_with_code(
+                UCode::RESOURCE_EXHAUSTED,
+                format!(
+                    "Maximum number of listeners reached: {}",
+                    self.max_subscribers
+                ),
+            ));
+        }
+
+        // Setup callback
+        let listener_cloned = comparable_listener.clone();
+        let callback = move |sample: Sample| {
+            // Get the UAttribute from Zenoh user attachment
+            let Some(attachment) = sample.attachment() else {
+                warn!(
+                    "Ignoring Zenoh Sample without attachment [key expr: {}]",
+                    sample.key_expr()
+                );
+                return;
+            };
+            let attributes = match attachment_to_uattributes(attachment) {
+                Ok(attribs) => attribs,
+                Err(e) => {
+                    warn!("Unable to transform attachment to UAttributes: {e:?}");
+                    return;
+                }
+            };
+            // we must only process messages that have not expired yet
+            // [impl->dsn~up-attributes-ttl~1]
+            // [impl->dsn~up-attributes-ttl-timeout~1]
+            if attributes.check_expired().is_ok() {
+                // Create UMessage
+                let msg = UMessage {
+                    attributes: Some(attributes).into(),
+                    payload: Some(sample.payload().to_bytes().to_vec().into()),
+                    ..Default::default()
+                };
+                spawn_nonblock_callback(&listener_cloned, msg);
+            } else if enabled!(Level::DEBUG) {
+                if let Some(id) = attributes.id.as_ref() {
+                    if let (Some(ts), Some(ttl)) = (id.get_time(), attributes.ttl) {
+                        debug!(
+                            "discarding expired message [id: {}, created: {ts}, ttl: {ttl}]",
+                            id.to_hyphenated_string(),
+                        );
+                    }
+                }
+            }
+        };
+
+        // Create Zenoh subscriber
+        // [impl->dsn~supported-message-delivery-methods~1]
+        match self
+            .session
+            // [impl->dsn~utransport-registerlistener-start-invoking-listeners~1]
+            .declare_subscriber(&zenoh_key)
+            .callback_mut(callback)
+            .await
+        {
+            Ok(subscriber) => {
+                // [impl->dsn~utransport-registerlistener-number-of-listeners~1]
+                // [impl->dsn~utransport-registerlistener-listener-reuse~1]
+                locked_subscribers.insert((zenoh_key, comparable_listener), subscriber);
+                Ok(())
+            }
+            Err(e) => {
+                let msg = "Failed to register listener";
+                info!("{msg}: {e}");
```

Comment:
We can use `warn` or` error` here.

Reply:
I am always hesitant to use these high log levels because in most cases, the situation we have run into is actually totally expected and handled properly and we only want to _duly note_ (in the log) that this has actually happened.
In this case I wonder if not being able to declare the subscriber on the Zenoh session is actually an expected case or indicates a (fundamental) error with the Zenoh setup that we cannot recover from, i.e. which requires restarting the system or some other (manual?) resolving?

Reply:
It's about how we define the log level. IMHO, when users call the function, they expect it to work. The warning messages (which have a different color) can easily help users identify why it fails.